src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -285,101 +307,126 @@ object RuleUtils {
+        // in formats other than "parquet" using one FileScan node as the operator requires
+        // files in one homogenous format. To address this, we need to read the appended
+        // source files using another FileScan node injected into the plan and subsequently
+        // merge the data into the index data. Please refer below [[Union]] operation.
+        // In case there are both deleted and appended files, we cannot handle the appended
+        // files along with deleted files as source files do not have the lineage column which
+        // is required for excluding the index data from deleted files.
+        // If the source relation is partitioned, we cannot read the appended files with the
+        // index data as the schema of partitioned files are not equivalent to the index data.
+        (index.content.files, filesAppended)
+      } else {
+        // If BucketSpec of index data isn't used (e.g., in the case of FilterIndex currently)
+        // and the source format is parquet, we could read the appended files along
+        // with the index data.
+        (index.content.files ++ filesAppended, Seq.empty)
+      }
+    }
+
+    def getIndexHadoopFsRelation(indexLogEntry: IndexLogEntry, plan: LogicalPlan,
+        filesDeleted: Set[FileInfo], filesToRead: Seq[Path]) = {
+      // In order to handle deleted files, read index data with the lineage column so that
+      // we could inject Filter-Not-In conditions on the lineage column to exclude the indexed
+      // rows from the deleted files.
+      val newSchema = StructType(
+        indexLogEntry.schema.filter(s =>
+          plan.schema.contains(s) || (filesDeleted.nonEmpty && s.name.equals(
+            IndexConstants.DATA_FILE_NAME_ID))))
+
+      val newLocation = new InMemoryFileIndex(spark, filesToRead, Map(), None)
+      HadoopFsRelation(
+        newLocation,
+        new StructType(),
+        newSchema,
+        if (useBucketSpec) Some(indexLogEntry.bucketSpec) else None,
+        new ParquetFileFormat,
+        Map(IndexConstants.INDEX_RELATION_IDENTIFIER))(spark)
+    }
+
+    def injectFilterNotForDeletedFiles(files: Set[FileInfo], fsRelation: HadoopFsRelation,
+        output: Seq[AttributeReference],
+        createRelationFn: (HadoopFsRelation, Seq[AttributeReference]) => LogicalRelation) = {
+      val updatedOutput =
+        output.filter(attr => fsRelation.schema.fieldNames.contains(attr.name))
+      if (files.isEmpty) {
+        createRelationFn(fsRelation, updatedOutput)
+      } else {
+        val lineageAttr = AttributeReference(IndexConstants.DATA_FILE_NAME_ID, LongType)()
+        val deletedFileIds = files.map(f => Literal(f.id)).toArray
+        val rel = createRelationFn(fsRelation, updatedOutput ++ Seq(lineageAttr))
+        val filterForDeleted = Filter(Not(In(lineageAttr, deletedFileIds)), rel)
+        Project(updatedOutput, OptimizeIn(filterForDeleted))
+      }
+    }
```

Comment:
`getIndexHadoopFsRelation`'s methods body is extracted from [L353-L382](https://github.com/microsoft/hyperspace/pull/321/files#diff-8c312528287b51f613809d58a7b14b23c151d65794c099b653605795e0a738edL353-L382).